sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/SchemaEvolutionTest.scala

```diff
@@ -0,0 +1,436 @@
+      val dir2 = s"$path${File.separator}part=three"
+
+      df1.write.mode("overwrite").format(format).options(options).save(dir1)
+      df2.write.mode("overwrite").format(format).options(options).save(dir2)
+
+      val df = spark.read
+        .schema(df1.schema)
+        .format(format)
+        .options(options)
+        .load(path)
+
+      checkAnswer(df, Seq(
+        Row("1", "a", "two"),
+        Row("2", "b", "two"),
+        Row("1", "a", "three"),
+        Row("2", "b", "three")))
+    }
+  }
+}
+
+/**
+ * Change column positions.
+ * This suite assumes that all data set have the same number of columns.
+ */
+trait ChangePositionEvolutionTest extends SchemaEvolutionTest {
+  import testImplicits._
+
+  test("change column position") {
+    withTempDir { dir =>
+      // val path = dir.getCanonicalPath
+      val path = "/tmp/change"
+
+      val df1 = Seq(("1", "a"), ("2", "b"), ("3", "c")).toDF("col1", "col2")
+      val df2 = Seq(("d", "4"), ("e", "5"), ("f", "6")).toDF("col2", "col1")
+      val unionDF = df1.unionByName(df2)
+
+      val dir1 = s"$path${File.separator}part=one"
+      val dir2 = s"$path${File.separator}part=two"
+
+      df1.write.mode("overwrite").format(format).options(options).save(dir1)
+      df2.write.mode("overwrite").format(format).options(options).save(dir2)
+
+      val df = spark.read
+        .schema(unionDF.schema)
+        .format(format)
+        .options(options)
+        .load(path)
+        .select("col1", "col2")
+
+      checkAnswer(df, unionDF)
+    }
+  }
+}
+
+trait BooleanTypeEvolutionTest extends SchemaEvolutionTest {
+  import testImplicits._
+
+  test("boolean to byte/short/int/long") {
+    withTempDir { dir =>
+      val path = dir.getCanonicalPath
+
+      val values = (1 to 10).map(_ % 2)
+      val booleanDF = (1 to 10).map(_ % 2 == 1).toDF("col1")
+      val byteDF = values.map(_.toByte).toDF("col1")
+      val shortDF = values.map(_.toShort).toDF("col1")
+      val intDF = values.toDF("col1")
+      val longDF = values.map(_.toLong).toDF("col1")
+
+      booleanDF.write.mode("overwrite").format(format).options(options).save(path)
+
+      val df1 = spark.read
+        .schema("col1 byte")
+        .format(format)
+        .options(options)
+        .load(path)
+      checkAnswer(df1, byteDF)
+
+      val df2 = spark.read
+        .schema("col1 short")
+        .format(format)
+        .options(options)
+        .load(path)
+      checkAnswer(df2, shortDF)
+
+      val df3 = spark.read
+        .schema("col1 int")
+        .format(format)
+        .options(options)
+        .load(path)
+      checkAnswer(df3, intDF)
+
+      val df4 = spark.read
+        .schema("col1 long")
+        .format(format)
+        .options(options)
+        .load(path)
+      checkAnswer(df4, longDF)
+    }
+  }
+}
+
+trait IntegralTypeEvolutionTest extends SchemaEvolutionTest {
+
+  import testImplicits._
+
+  test("change column type from `byte` to `short/int/long`") {
```

Comment:
nit: `` `byte` `` -> `byte` or the opposite for consistency with the same instances.